### ISSUE TITLE
`tf.nn.silu` have wrong gradient for complex input

### ISSUE THREAD

**Issue by VictoriaGriffith** [tensorflow/tensorflow#57357]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

source

### Tensorflow Version

tf 2.9

### Custom Code

Yes

### OS Platform and Distribution

Linux Ubuntu 20.04

### Python version

3.9

### Current Behaviour?

```shell
`tf.nn.silu(x)` is equal to `x * sigmoid(x)`, however, when I compute the gradient with respect to a complex tensor, the gradient of the two functions are different. I believe that the `tf.nn.silu`-gradient is wrong because it is inconsistent with the numerical gradient.
```

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
input = tf.complex([[[0.0], [0.0]]], [[[5.35], [6.61]]])
print("input: ", input)
with tf.GradientTape() as t:
  t.watch(input)
  output1 = tf.nn.silu(input)
grad1 = t.gradient(output1, input)
with tf.GradientTape() as t:
  t.watch(input)
  output2 = input * tf.math.sigmoid(input)
grad2 = t.gradient(output2, input)

assert tf.experimental.numpy.allclose(output1, output2) # Assertion Pass
print(grad1)
print(grad2)
assert tf.experimental.numpy.allclose(grad1, grad2) # AssertionError
```

### Relevant log output

```shell
input:  tf.Tensor(
[[[0.+5.35j]
  [0.+6.61j]]], shape=(1, 2, 1), dtype=complex64)
tf.Tensor(
[[[0.5+1.4249809j]
  [0.5+1.7798613j]]], shape=(1, 2, 1), dtype=complex64)
tf.Tensor(
[[[0.5-1.4249809j]
  [0.5-1.7798615j]]], shape=(1, 2, 1), dtype=complex64)
AssertionError
```
</details>

**Comment by VictoriaGriffith** [comment]
Here is the theoretical and numerical gradient for `tf.nn.silu`:
```
import tensorflow as tf
input = tf.complex([[[0.0], [0.0]]], [[[5.35], [6.61]]])
print("input: ", input)
output1 = tf.nn.silu(input)
theoretical1, numerical1 = tf.test.compute_gradient(tf.nn.silu, [input], delta=1e-06)
print("theoretical1: ", theoretical1)
print("numerical1: ", numerical1)
from tensorflow.python.ops import gradient_checker_v2
max_error = gradient_checker_v2.max_error(theoretical1, numerical1)
print("max_error with tf.nn.silu: ", max_error)

def f(x):
  return x * tf.math.sigmoid(x)
output2 = f(input)
assert tf.experimental.numpy.allclose(output1, output2)

theoretical2, numerical2 = tf.test.compute_gradient(f, [input], delta=1e-06)
max_error = gradient_checker_v2.max_error(theoretical2, numerical2)
print("max_error with x * sigmoid(x): ", max_error)
max_error = gradient_checker_v2.max_error(theoretical2, numerical1)
print("max_error with numerical1 and theoretical2:", max_error)
```
Outputs:
```
input:  tf.Tensor(
[[[0.+5.35j]
  [0.+6.61j]]], shape=(1, 2, 1), dtype=complex64)
theoretical1:  (array([[ 0.5      ,  1.4249809,  0.       ,  0.       ],
       [-1.4249809,  0.5      ,  0.       ,  0.       ],
       [ 0.       ,  0.       ,  0.5      ,  1.7798613],
       [ 0.       ,  0.       , -1.7798613,  0.5      ]], dtype=float32),)
numerical1:  (array([[ 0.5364418 , -1.4305115 ,  0.        ,  0.        ],
       [ 1.5497208 ,  0.47683716,  0.        ,  0.        ],
       [ 0.        ,  0.        ,  0.44703484, -1.66893   ],
       [ 0.        ,  0.        ,  1.7881393 ,  0.47683716]],
      dtype=float32),)
max_error with tf.nn.silu:  3.5680007934570312
max_error with x * sigmoid(x):  0.1247398853302002
max_error with numerical1 and theoretical2: 0.1247398853302002
```

**Comment by VictoriaGriffith** [comment]
I check `sigmoid` and find similar bug also exists in `tf.nn.sigmoid`:
```
import tensorflow as tf
x = tf.complex([[0.0, 0.0], [0, 0]], [[9.4, 12], [1, 6]])
print(x)
theoretical1, numerical1 = tf.test.compute_gradient(tf.nn.sigmoid, [x], delta=1e-06)

from tensorflow.python.ops import gradient_checker_v2
max_error = gradient_checker_v2.max_error(theoretical1, numerical1)
print("max_error with tf.nn.sigmoid: ", max_error)
```
Outputs:
```
tf.Tensor(
[[0. +9.4j 0.+12.j ]
 [0. +1.j  0. +6.j ]], shape=(2, 2), dtype=complex64)
max_error with tf.nn.sigmoid:  76.2554931640625
```

**Comment by sushreebarsa** [comment]
@VictoriaGriffith I tried to replicate the issue on colab, could you please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/c5115d5cb970ea9ce138d2e49a34e981/57357.ipynb) and confirm the same?
Thank you!

**Comment by VictoriaGriffith** [comment]
Yes, you're right, the `allclose` assertion for outputs pass, but the `allclose` assertion for gradients fail, indicating a bug in gradient computation. Thanks for replicating the issue @sushreebarsa!

**Comment by cantonios** [comment]
I double-checked analytically for sigmoid, and the theoritical gradient is correct (similarly for silu).

The issue is that for pure complex numbers, there is a numerical singularity at `pi + 2 * k *pi` for integers `k` (i.e. odd multiples of `pi`).  Your 9.4 is very close to this, so the numerical gradient starts to lose accuracy.  IOW, it's the numerical gradient here that is poorly conditioned around 9.4, causing the deviation.

Casting to `complex128` reduces the error in the numerical gradient significantly.
